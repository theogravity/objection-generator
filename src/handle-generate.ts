import Ajv from 'ajv'
import { resolve, dirname } from 'path'

import { Arguments } from './interfaces'
import { dereference, loadYaml } from './utils'
import { validateConfig } from './schema-validator'
import { ObjectionGenerator } from './generators/objection'
import { KnexGenerator } from './generators/knex'
import { BaseGenerator } from './generators/base'

export async function handleGenerateObjection (params: Arguments) {
  params.templateDir = params.templateDir
    ? params.templateDir
    : resolve(__dirname, 'templates', 'objection')

  await handleGenerate(ObjectionGenerator, params)

  console.log('Models generated in:', params.outDir)
}

export async function handleGenerateKnex (params: Arguments) {
  params.templateDir = params.templateDir
    ? params.templateDir
    : resolve(__dirname, 'templates', 'knex')

  await handleGenerate(KnexGenerator, params)
  console.log('Migration generated in:', params.outDir)
}

export async function handleGenerate (
  Generator: typeof BaseGenerator,
  params: Arguments
) {
  try {
    const specPath = resolve(process.cwd(), params.specFile)
    const outDir = resolve(process.cwd(), params.outDir)

    const data = await loadYaml(specPath)

    console.log('Validating YAML spec')

    await validateConfig(data)

    console.log('Generating')

    const dereferenced = await dereference(data, dirname(specPath))

    const generator = new Generator(dereferenced, params.templateDir, outDir)
    await generator.init()
    await generator.generate()
  } catch (err) {
    if (err instanceof Ajv.ValidationError) {
      console.log(JSON.stringify(err.errors, null, 2))
      process.exit(1)
    }

    console.log(err.message)
    process.exit(1)
  }
}
