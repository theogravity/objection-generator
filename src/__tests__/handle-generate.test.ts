/* eslint-env jest */

import { pathExistsSync, remove } from 'fs-extra'
import { readFileSync } from 'fs'
import { join } from 'path'

import { handleGenerate } from '../handle-generate'

const OUT_DIR = '/tmp/generator-models'

beforeEach(async () => {
  await remove(OUT_DIR)
})

describe('handle-generator', () => {
  it('should generate models', async () => {
    await handleGenerate({
      specFile: join(__dirname, '..', '..', 'sample.yaml'),
      outDir: OUT_DIR
    })

    const movieModelPath = join(OUT_DIR, 'MovieModel.ts')

    expect(pathExistsSync(movieModelPath)).toBe(true)
    expect(readFileSync(movieModelPath, 'utf8')).toMatchSnapshot()
  })
})
