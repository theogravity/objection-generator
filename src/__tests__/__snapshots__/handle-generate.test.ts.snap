// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handle-generator should generate a knex migration 1`] = `
"async function up (knex) {
  await knex.schema.createTable('persons', table => {
    table.string('id')
    table.string('name', 100).notNullable()
    table.integer('age')
    table.enu('gender', ['Male', 'Female', 'Other']).defaultTo('Female')
    table
      .string('username', 25)
      .notNullable()
      .defaultTo('default-user')
    table.datetime('created').defaultTo(knex.fn.now())

    table.primary(['id'])

    table.unique(['username'], 'uniq_username')

    table.index(['age', 'name'], 'name_age_index')
  })
  await knex.schema.createTable('movies', table => {
    table.string('id')
    table.string('name', 255).notNullable()

    table.primary(['id'])
  })
  await knex.schema.createTable('reviews', table => {
    table.string('review_id')
    table.string('author_id').notNullable()
    table.string('movie_id').notNullable()
    table.string('content')

    table.primary(['review_id'])
  })
}

async function down (knex) {
  await knex.schema.dropTable('persons')
  await knex.schema.dropTable('movies')
  await knex.schema.dropTable('reviews')
}

module.exports = {
  up,
  down
}
"
`;

exports[`handle-generator should generate objection models 1`] = `
"import path from 'path'

import { BaseModel } from './BaseModel'

export class MovieModel extends BaseModel {
  id: string
  name: string

  static tableName = 'movies'

  static get jsonSchema () {
    return {
      type: 'object',
      required: ['name'],
      properties: {
        id: {
          type: 'string'
        },
        name: {
          type: 'string',
          minLength: 1,
          maxLength: 255
        }
      }
    }
  }

  static get relationMappings () {
    return {
      reviews: {
        relation: Model.HasManyRelation,
        modelClass: path.join(__dirname, 'ReviewModel'),
        join: {
          from: 'movie.id',
          to: 'review.movieId'
        }
      }
    }
  }
}
"
`;

exports[`handle-generator should generate objection models 2`] = `
"import path from 'path'

import { BaseModel } from './BaseModel'

export class ReviewModel extends BaseModel {
  reviewId: string
  authorId: string
  movieId: string
  content: string

  static tableName = 'reviews'

  static get jsonSchema () {
    return {
      type: 'object',
      required: ['authorId', 'movieId'],
      properties: {
        reviewId: {
          type: 'string'
        },
        authorId: {
          type: 'string'
        },
        movieId: {
          type: 'string'
        },
        content: {
          type: 'string'
        }
      }
    }
  }

  static get idColumn () {
    return 'reviewId'
  }

  static get relationMappings () {
    return {
      author: {
        relation: Model.HasOneRelation,
        modelClass: path.join(__dirname, 'PersonModel'),
        join: {
          from: 'reviews.authorId',
          to: 'persons.id'
        }
      }
    }
  }
}
"
`;
